src/metabase/query_processor/middleware/update_used_cards.clj

```diff
@@ -12,14 +13,25 @@
 (set! *warn-on-reflection* true)
 
 (defn- update-used-cards!*
-  [used-card-ids]
-  (when (seq used-card-ids)
-    (qp.util/with-execute-async
-      (fn []
-        (try
-          (t2/update! :model/Card :id [:in used-card-ids] {:last_used_at :%now})
-          (catch Throwable e
-            (log/error e "Error updating used cards")))))))
+  [card-id-timestamps]
+  (let [card-id->timestamp (update-vals (group-by :id card-id-timestamps)
+                                        (fn [xs] (apply t/max (map :timestamp xs))))]
+    (log/debugf "Update last_used_at of %d cards" (count card-id->timestamp))
+    (try
+      (t2/update! :model/Card :id [:in (keys card-id->timestamp)]
+                  {:last_used_at (into [:case]
+                                       (apply concat (for [[id timestamp] card-id->timestamp]
+                                                       [[:= :id id] [:greatest :last_used_at timestamp]])))})
```

Comment:
I don't think this solves Dan's concern, which was around the latency until this db update occurs, during which the cards will not be seen as recently used.

I don't see an easier way of solving that having an in-memory map that is updated when we add to the queue, and partially cleared whenever we process a batch. As for cross-instance complexity I think it's fair for such deployments to experience less consistency. As we move to an indexing approach for search as well, some lag in navigation will be unavoidable.

An even simpler mitigation is to decrease the interval as well. How did you arrive at the interval you are using? When you extract the def(s) for the interval(s) it would be useful to record the line of thinking in the docstring.

Reply:
Regarding interval: 5 mins seem like a really long time to me, I'd prefer something in second to tens of seconds, but I don't know what kind of load it generates in reality.

Reply:
yea, 5 minutes is too much. how about 20 seconds?

Reply:
Agree, that shouldn't put excessive load and is short enough to to be extremely irritating (at the first look, not sure how it will turn out for real people, haha).